### PR TITLE
#617 unique planes per inline model

### PR DESCRIPTION
**Pull request by jdolan**
@Paril This is in a branch because I'm a little worried about regression.

The fix to ensure we have unique planes for each BSP inline model is the one-liner in map.c to `memset` the plane hash. But I immediately noticed that maps failed to load with this change, because brush sides were referencing invalid planes. How could that be?

Turns out, there are two instances of brush beveling code, for some reason. One instance is in map.c, AddBrushBevels, the function we all know and love (I hate that code and have yet to make it pretty). But through debugging this, I realized that there's another spot where the code adds bevels: when it is emitting the brushes into the BSP file!

This happens very, very late in the process. In fact, this is even **after** planes have already been written to disk. Any new planes created by this bevel routine would not be included in the BSP. For all of these years, we've just apparently gotten lucky that no late-bevels needed to create new planes. I'm not sure there's anything mathematically guaranteeing that would be the case..

So I decided to tack that late-stage bevel code onto the map.c bevel code. This appears to work as expected, but.. anytime we're dealing with BSP format and collision model, I've learned to be cautious. So here it is, for your review.

### CODE REVIEW

File: src/tools/quemap/map.c [hunk 1]
```diff
@@ -769,6 +808,8 @@ static entity_t *ParseEntity(parser_t *parser) {
 			Com_Error(ERROR_FATAL, "MAX_BSP_ENTITIES\n");
 		}
 
+		memset(plane_hash, 0, sizeof(plane_hash));
+
```
Comment (jdolan) [inline src/tools/quemap/map.c:812]: Were it not for the bevel nonsense, this line alone would have been enough to guarantee unique planes for each inline model, heh.

File: src/tools/quemap/writebsp.c [hunk 2]
```diff
@@ -29,19 +29,21 @@
...
-		out->dist = (float) p->dist;
+		out->normal = in->normal;
+		out->dist = (float) in->dist;
 
 		bsp_file.num_planes++;
 
 		Progress("Emitting planes", 100.f * i / num_planes);
 	}
+
+	return bsp_file.num_planes;
```
Comment (jdolan) [inline src/tools/quemap/writebsp.c:46]: Purely cosmetic.

File: src/tools/quemap/writebsp.c [hunk 3]
```diff
@@ -376,42 +378,6 @@ static void EmitBrushes(void) {
...
-				normal.xyz[axis] = side;
-
-				float dist;
-				if (side == -1) {
-					dist = -b->bounds.mins.xyz[axis];
-				} else {
-					dist = b->bounds.maxs.xyz[axis];
-				}
-
-				const int32_t plane_num = FindPlane(normal, dist);
```
Comment (jdolan) [inline src/tools/quemap/writebsp.c:393]: So here is the late-stage beveling code. This way too late to be doing this, in my opinion. The BSP is finalized at this point, and we're just flushing it to disk. FindPlane can and does create new planes if necessary, and in this instance, those planes would never make it to the BSP. So I thought it best to relocate this code to AddBrushBevels.

**Comment by Paril** [comment]
I dunno if the plan is to touch AddBrushBevels here, but Q3A did a little adjustment to brush bevels: they ensured that the first six brush sides/planes were always axial, which let them skip some checks in the tracing code. While it likely won't matter for most cases, it might help the new rotation code perhaps since it's a bit heavier?

I recompiled all of the maps locally, going to test them all out and see if I notice anything off. I notice no warnings while compiling (except for 2 benign "entity outside of map" warnings on Tokays, and this one in Warehouse:
```
WARNING: BrushContents: Entity 0, Brush 167: Mixed face contents: 1 expected 1000000
worldspawn @ (0.00 0.00 0.00)
```
)

**Comment by Paril** [comment]
image
Edge has some black spots here

**Comment by Paril** [comment]
image
Some weird lighting here, not sure if this was like this before or not. Just -bsp -light at command line.

**Comment by Paril** [comment]
image
image
image

Fractures. On last image, left has texture misalignments, right has dark faces

**Comment by Paril** [comment]
image
https://streamable.com/bznctv lighting on bmodels here seems all wrong, but I can't put my finger on it.
image
image

**Comment by Paril** [comment]
image

**Comment by Paril** [comment]
image
I think this should be enough info to look into the bugs. Code-wise it looks solid but there's still somethin that isn't working properly.

**Comment by Paril** [comment]
image
Just noticed this one by chance while testing a different bugfix

**Comment by jdolan** [comment]
The black faces and broken planes are a consequence of fog entities getting unique planes, but then being merged into worldspawn. This creates duplicate planes which conflict with each other during BSP tree creation.

Fixing this will be more involved. Since this is a low priority, affecting just one map, and with an easy workaround fix, I'm going to abandon this attempt, to possibly revisit it later. I thought it would be an easy win, but.. as with most things in the BSP compiler, nope.

**Pull request closed without merging**